test/http-request.ts

```diff
@@ -0,0 +1,332 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as assert from 'assert';
+import {describe, it, beforeEach} from 'mocha';
+import {ServerResponse} from 'http';
+import {
+  makeHttpRequestData,
+  ServerRequest,
+  makeHeaderWrapper,
+  getCloudTraceContext,
+  getTraceContext,
+} from '../src/http-request';
+import * as http from 'http';
+import * as proxyquire from 'proxyquire';
+
+describe('format raw http request to structured http-request', () => {
+  describe('makeHttpRequestData', () => {
+    it('should prioritize originalUrl if provided', () => {
+      const req = {
+        method: 'GET',
+        url: 'http://wrongemail.com/',
+        originalUrl: 'http://google.com/',
+      } as ServerRequest;
+      const cloudReq = makeHttpRequestData(req);
+      assert.strictEqual(cloudReq.protocol, 'http:');
+      assert.strictEqual(cloudReq.requestUrl, 'http://google.com/');
+      assert.strictEqual(cloudReq.requestMethod, 'GET');
+    });
+    it('should infer as many request values as possible', () => {
+      const req = {
+        method: 'GET',
+        url: 'http://google.com/',
+        headers: {
+          'user-agent': 'some-agent',
+          referer: 'some-referer',
+        },
+      } as http.IncomingMessage;
+      const cloudReq = makeHttpRequestData(req);
+      assert.strictEqual(cloudReq.protocol, 'http:');
+      assert.strictEqual(cloudReq.requestUrl, 'http://google.com/');
+      assert.strictEqual(cloudReq.requestMethod, 'GET');
+      assert.strictEqual(cloudReq.userAgent, 'some-agent');
+      assert.strictEqual(cloudReq.referer, 'some-referer');
+      assert.strictEqual(cloudReq.status, undefined);
+    });
+    it('should infer header values with caution', () => {
+      const req = {
+        method: 'GET',
+        url: 'http://google.com/',
+        headers: {
+          'user-agent': '',
+          referer: undefined,
+        },
+      } as http.IncomingMessage;
+      const cloudReq = makeHttpRequestData(req);
+      assert.strictEqual(cloudReq.protocol, 'http:');
+      assert.strictEqual(cloudReq.requestUrl, 'http://google.com/');
+      assert.strictEqual(cloudReq.requestMethod, 'GET');
+      assert.strictEqual(cloudReq.userAgent, undefined);
+      assert.strictEqual(cloudReq.referer, undefined);
+      assert.strictEqual(cloudReq.status, undefined);
+    });
+    it('should infer as many response values as possible', () => {
+      const RESPONSE_SIZE = 2048;
```

Comment:
size in what units?

Reply:
it just mocks the content length header, e.g. `content-length: 6553`, there's no unit that needs to be specified.